File: 63.不同路径-ii.c
/*
 * @lc app=leetcode.cn id=63 lang=c
 *
 * [63] 不同路径 II
 */

// @lc code=start

int uniquePathsWithObstacles(int **obstacleGrid, int obstacleGridSize, int *obstacleGridColSize)
{
  int **dp, res, i, j;
  res = 0;
  dp = (int **)malloc(sizeof(int *) * obstacleGridSize);
  for (i = 0; i < obstacleGridSize; i++)
    dp[i] = (int *)malloc(sizeof(int) * (*obstacleGridColSize));

  for (i = 0; i < obstacleGridSize; i++)
    dp[i][0] = 1;
  for (i = 0; i < *obstacleGridColSize; i++)
    dp[0][i] = 1;
  for (i = 0; i < obstacleGridSize; i++)
  {
    if (obstacleGrid[i][0] == 1)
    {
      for (j = i; j < obstacleGridSize; j++)
        dp[j][0] = 0;
      break;
    }
  }
  for (i = 0; i < *obstacleGridColSize; i++)
  {
    if (obstacleGrid[0][i] == 1)
    {
      for (j = i; j < *obstacleGridColSize; j++)
        dp[0][j] = 0;
      break;
    }
  }

  for (i = 0; i < obstacleGridSize; i++)
    for (j = 0; j < *obstacleGridColSize; j++)
      if (obstacleGrid[i][j] == 1)
        dp[i][j] = 0;

  for (i = 1; i < obstacleGridSize; i++)
  {
    for (j = 1; j < *obstacleGridColSize; j++)
    {
      if (dp[i][j] == 0)
        continue;
      dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
    }
  }
  return dp[obstacleGridSize - 1][*obstacleGridColSize - 1];
}
// @lc code=end

File: store/63.不同路径-ii.js
/*
 * @lc app=leetcode.cn id=63 lang=javascript
 *
 * [63] 不同路径 II
 */

// @lc code=start
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function (obstacleGrid) {
  let dp = new Array(obstacleGrid.length)
    .fill("")
    .map(() => Array(obstacleGrid[0].length).fill(0));

  for (let i = 0; i < dp.length; i++) {
    if (obstacleGrid[i][0]) break;
    dp[i][0] = 1;
  }

  for (let j = 0; j < dp[0].length; j++) {
    if (obstacleGrid[0][j]) break;
    dp[0][j] = 1;
  }

  for (let i = 1; i < dp.length; i++) {
    for (let j = 1; j < dp[0].length; j++) {
      if (obstacleGrid[i][j]) dp[i][j] = 0;
      else dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
    }
  }

  return dp[dp.length - 1][dp[0].length - 1];
};
// @lc code=end
